models/base.py

```diff
@@ -179,6 +183,8 @@ def __call__(self, *args, name: Optional[str] = None, **kwargs) -> Layer:
           layer_dict["reuse_params"] = self.calls[0].get_name()
       layer = Layer(self, layer_dict)
       self.calls.append(layer)
+      if ret:
+        return (layer, *ret)
```

Comment:
The return is wrong. Also, the brackets are not needed (and thus should not be used).
```suggestion
        return layer, ret
```
Also, the type annotation of the function (`__call__`) should also be changed.

Reply:
Yes I realize the typing is wrong, I did not think about that yet, but I feel like this is something we can fix once the code is "done"
The brackets where added for one of the test i think, since they test in python 3.7 iirc, and from what I understand, returning without brackets is only possible since 3.8. But maybe I am wrong on this part. 
If we only do `return ret` then we dont return the layer that is intended to be the output of the subnetwork, since `ret = layer_dict[1:]` or am I missing something?

Reply:
> Yes I realize the typing is wrong, I did not think about that yet, but I feel like this is something we can fix once the code is "done"

Better do it directly now that we already have it here and do not forget about it.

> The brackets where added for one of the test i think, since they test in python 3.7 iirc, and from what I understand, returning without brackets is only possible since 3.8. But maybe I am wrong on this part.

Returning without brackets was possible since Python 1 I think.

> If we only do `return ret` then we dont return the layer that is intended to be the output of the subnetwork, since `ret = layer_dict[1:]` or am I missing something?

Yes exactly. We instead would return the same structure which is intended to be returned, exactly what `forward` returned, i.e. a tuple of layers, or whatever structure `forward` returned.

Reply:
>Returning without brackets was possible since Python 1 I think.

Hmm there was an error with one commit without these brackets, I think it was related to syntax including * so e.g. `layer, *ret` I think that failed in one of the tests, but we removed that anyways

>Yes exactly. We instead would return the same structure which is intended to be returned, exactly what forward returned, i.e. a tuple of layers, or whatever structure forward returned.

Ahh I think I understand: So because in the `make_layer_dict` function we would return as the second argument all of the layers passed, this would here then mean that the output is also in there so we would not lose any reference to any layer right? So if `forward` returned `out, (hidden, last)` then `make_layer_dict` returns `LayerDict, (out, (hidden, last))` right? Then it makes sense to me.

Reply:
> Hmm there was an error with one commit without these brackets, I think it was related to syntax including * so e.g. `layer, *ret` I think that failed in one of the tests

Ah yes, that's possible.

> Ahh I think I understand: So because in the `make_layer_dict` function we would return as the second argument all of the layers passed, this would here then mean that the output is also in there so we would not lose any reference to any layer right?

Yes. In the first case with only one layer, this becomes the "output" of the subnetwork, and then the subnetwork itself is fine as a layer-ref (which doesn't need to be explicitly specified because it gets the layer dict).

In the case with multiple outputs / some nested structure, we just pass on all layer refs as-is. The "output" of the subnetwork, and the subnetwork layer itself would actually not directly be used but just the layer refs directly (e.g. like "subnet/out", "subnet/hidden", "subnet/last", whatever).

> So if `forward` returned `out, (hidden, last)` then `make_layer_dict` returns `LayerDict, (out, (hidden, last))` right? Then it makes sense to me.

Yes exactly. And this `ILayerMaker.__call__` would return again only the `out, (hidden, last)`.